src/renderer/components/shared/CanvasDataBuilder.js

```diff
@@ -123,17 +123,56 @@ const getInstanceNode = async (instance, graqlVar, explanation) => {
   return node;
 };
 
+/**
+ * produces the `has` edges from owners to the given attribute instance
+ * @param {Concept} attribute must be an attribute instance
+ */
+const getInstanceHasEdges = async (attribute) => {
+  const owners = (await (await attribute.owners()).collect());
+  const edges = await Promise.all(owners.map(owner => getEdge(owner, attribute, edgeTypes.instance.HAS)));
+  return edges;
+};
+
+/**
+ * produces the `role` edges from the given relation instance to its roleplayers
+ * @param {Concept} relation must be a relation instance
+ */
+const getInstanceRoleEdges = async (relation) => {
+  const edges = [];
+  const rpMap = await relation.rolePlayersMap();
+  const roleAndPlayersMap = Array.from(rpMap.entries());
+  // eslint-disable-next-line no-restricted-syntax
+  for (const roleAndPlayers of roleAndPlayersMap) {
+    const role = roleAndPlayers[0];
+    const players = roleAndPlayers[1];
+
+    const roleLabel = await role.label();
+    // eslint-disable-next-line no-loop-func
+    players.forEach(player => edges.push({ to: player.id, label: roleLabel, from: relation.id }));
+  }
```

Comment:
minor, but this code could be:
```
...
rpMap.forEach(players, role => {
    const roleLabel = await role.label();
    // eslint-disable-next-line no-loop-func
    players.forEach(player => edges.push({ to: player.id, label: roleLabel, from: relation.id }))
})
```

Reply:
can't do `await` in `forEach` :/ 

noticed another problem with this code though; we should always leave the construction of the edge object to `getEdge()`, which isn't the case on line 151

Reply:
yeah, in general though we should find a way not to block in loops, i.e. avoid
```
 for (var of bananas){
    const ciao = await superlongRequestToServer(); // this blocks the whole thread
 }
```
so maybe something along the lines of :
```
const promizez = Array.from(rpMap).map(([players, role]) => {
    role.label().then((label) => getEdge(relation, player, edgeTypes.instance.relates, roleLabel))
})
const edgezzz = Promise.all(promizez).(flatMap here probably)
```

Reply:
I've made an issue for this: https://github.com/graknlabs/workbase/issues/206
I'll work on it on a separate PR as the change proposed here may need to be applied to other functions in the `CanvasDataBuilder`